File: sdk/rust/src/schemas/string.rs
use regex::Regex;
use serde_json::{json, Value};

use crate::format::validators::validate_format;
use crate::issue_codes::*;
use crate::schema::{ParseContext, Schema};
use crate::types::{PathSegment, ValidationIssue, value_type_name};

/// Schema for string validation with optional constraints.
#[derive(Debug, Clone)]
pub struct StringSchema {
    pub min_length: Option<usize>,
    pub max_length: Option<usize>,
    pub pattern: Option<String>,
    pub starts_with: Option<String>,
    pub ends_with: Option<String>,
    pub includes: Option<String>,
    pub format: Option<String>,
    pub coerce: Option<Vec<String>>,
    pub default_value: Option<Value>,
}

impl StringSchema {
    pub fn new() -> Self {
        StringSchema {
            min_length: None,
            max_length: None,
            pattern: None,
            starts_with: None,
            ends_with: None,
            includes: None,
            format: None,
            coerce: None,
            default_value: None,
        }
    }

    pub fn min_length(mut self, n: usize) -> Self {
        self.min_length = Some(n);
        self
    }

    pub fn max_length(mut self, n: usize) -> Self {
        self.max_length = Some(n);
        self
    }

    pub fn pattern(mut self, p: &str) -> Self {
        self.pattern = Some(p.to_string());
        self
    }

    pub fn starts_with(mut self, s: &str) -> Self {
        self.starts_with = Some(s.to_string());
        self
    }

    pub fn ends_with(mut self, s: &str) -> Self {
        self.ends_with = Some(s.to_string());
        self
    }

    pub fn includes(mut self, s: &str) -> Self {
        self.includes = Some(s.to_string());
        self
    }

    pub fn format(mut self, f: &str) -> Self {
        self.format = Some(f.to_string());
        self
    }

    pub fn coerce(mut self, c: Vec<String>) -> Self {
        self.coerce = Some(c);
        self
    }

    pub fn default(mut self, v: Value) -> Self {
        self.default_value = Some(v);
        self
    }
}

impl Default for StringSchema {
    fn default() -> Self {
        Self::new()
    }
}

impl Schema for StringSchema {
    fn kind(&self) -> &str {
        "string"
    }

    fn parse_value(
        &self,
        input: &Value,
        path: &[PathSegment],
        _ctx: &ParseContext,
    ) -> Result<Value, Vec<ValidationIssue>> {
        // Apply coercions
        let mut value = input.clone();
        if let Some(coercions) = &self.coerce {
            for c in coercions {
                match c.as_str() {
                    "trim" => {
                        if let Value::String(s) = &value {
                            value = Value::String(s.trim().to_string());
                        }
                    }
                    "lower" => {
                        if let Value::String(s) = &value {
                            value = Value::String(s.to_lowercase());
                        }
                    }
                    "upper" => {
                        if let Value::String(s) = &value {
                            value = Value::String(s.to_uppercase());
                        }
                    }
                    _ => {}
                }
            }
        }

        let s = match &value {
            Value::String(s) => s.clone(),
            other => {
                return Err(vec![ValidationIssue {
                    code: INVALID_TYPE.to_string(),
                    path: path.to_vec(),
                    expected: "string".to_string(),
                    received: value_type_name(other).to_string(),
                    meta: None,
                }]);
            }
        };

        let mut issues = Vec::new();

        if let Some(min) = self.min_length {
            if s.len() < min {
                issues.push(ValidationIssue {
                    code: TOO_SMALL.to_string(),
                    path: path.to_vec(),
                    expected: min.to_string(),
                    received: s.len().to_string(),
                    meta: None,
                });
            }
        }

        if let Some(max) = self.max_length {
            if s.len() > max {
                issues.push(ValidationIssue {
                    code: TOO_LARGE.to_string(),
                    path: path.to_vec(),
                    expected: max.to_string(),
                    received: s.len().to_string(),
                    meta: None,
                });
            }
        }

        if let Some(pat) = &self.pattern {
            if let Ok(re) = Regex::new(pat) {
                if !re.is_match(&s) {
                    issues.push(ValidationIssue {
                        code: INVALID_STRING.to_string(),
                        path: path.to_vec(),
                        expected: pat.clone(),
                        received: s.clone(),
                        meta: None,
                    });
                }
            }
        }

        if let Some(prefix) = &self.starts_with {
            if !s.starts_with(prefix.as_str()) {
                issues.push(ValidationIssue {
                    code: INVALID_STRING.to_string(),
                    path: path.to_vec(),
                    expected: prefix.clone(),
                    received: s.clone(),
                    meta: None,
                });
            }
        }

        if let Some(suffix) = &self.ends_with {
            if !s.ends_with(suffix.as_str()) {
                issues.push(ValidationIssue {
                    code: INVALID_STRING.to_string(),
                    path: path.to_vec(),
                    expected: suffix.clone(),
                    received: s.clone(),
                    meta: None,
                });
            }
        }

        if let Some(sub) = &self.includes {
            if !s.contains(sub.as_str()) {
                issues.push(ValidationIssue {
                    code: INVALID_STRING.to_string(),
                    path: path.to_vec(),
                    expected: sub.clone(),
                    received: s.clone(),
                    meta: None,
                });
            }
        }

        if let Some(fmt) = &self.format {
            if !validate_format(fmt, &s) {
                issues.push(ValidationIssue {
                    code: INVALID_STRING.to_string(),
                    path: path.to_vec(),
                    expected: fmt.clone(),
                    received: s.clone(),
                    meta: None,
                });
            }
        }

        if issues.is_empty() {
            Ok(value)
        } else {
            Err(issues)
        }
    }

    fn export_node(&self) -> Value {
        let mut node = json!({"kind": "string"});
        let obj = node.as_object_mut().unwrap();
        if let Some(v) = self.min_length {
            obj.insert("minLength".to_string(), json!(v));
        }
        if let Some(v) = self.max_length {
            obj.insert("maxLength".to_string(), json!(v));
        }
        if let Some(v) = &self.pattern {
            obj.insert("pattern".to_string(), json!(v));
        }
        if let Some(v) = &self.starts_with {
            obj.insert("startsWith".to_string(), json!(v));
        }
        if let Some(v) = &self.ends_with {
            obj.insert("endsWith".to_string(), json!(v));
        }
        if let Some(v) = &self.includes {
            obj.insert("includes".to_string(), json!(v));
        }
        if let Some(v) = &self.format {
            obj.insert("format".to_string(), json!(v));
        }
        if let Some(v) = &self.coerce {
            if v.len() == 1 {
                obj.insert("coerce".to_string(), json!(v[0]));
            } else {
                obj.insert("coerce".to_string(), json!(v));
            }
        }
        if let Some(v) = &self.default_value {
            obj.insert("default".to_string(), v.clone());
        }
        node
    }
}

File: sdk/js/src/schemas/string.ts
import type { ParseContext, SchemaNode, StringFormat } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";
import { validateFormat } from "../format/validators.js";
import { describeType } from "../util.js";

export class StringSchema extends BaseSchema<string, string> {
  private _minLength?: number;
  private _maxLength?: number;
  private _pattern?: string;
  private _startsWith?: string;
  private _endsWith?: string;
  private _includes?: string;
  private _format?: StringFormat;
  /**
   * Cached compiled pattern. `undefined` = not yet compiled, `null` =
   * compilation failed (invalid pattern). Avoids recompiling on every value.
   */
  private _patternRe?: RegExp | null;

  _getCoercionTarget(): string {
    return "string";
  }

  minLength(n: number): this {
    const clone = this._clone();
    clone._minLength = n;
    return clone;
  }

  maxLength(n: number): this {
    const clone = this._clone();
    clone._maxLength = n;
    return clone;
  }

  pattern(p: string): this {
    const clone = this._clone();
    clone._pattern = p;
    // Reset cached compilation inherited from the source via _clone().
    clone._patternRe = undefined;
    return clone;
  }

  /** Lazily compile and cache the pattern. Returns null if invalid. */
  private _getPatternRe(): RegExp | null {
    if (this._patternRe !== undefined) return this._patternRe;
    try {
      this._patternRe = new RegExp(this._pattern as string);
    } catch {
      this._patternRe = null;
    }
    return this._patternRe;
  }

  startsWith(s: string): this {
    const clone = this._clone();
    clone._startsWith = s;
    return clone;
  }

  endsWith(s: string): this {
    const clone = this._clone();
    clone._endsWith = s;
    return clone;
  }

  includes(s: string): this {
    const clone = this._clone();
    clone._includes = s;
    return clone;
  }

  format(f: StringFormat): this {
    const clone = this._clone();
    clone._format = f;
    return clone;
  }

  _validate(input: unknown, ctx: ParseContext): unknown {
    if (typeof input !== "string") {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_TYPE,
        message: `Expected string, received ${describeType(input)}`,
        path: [...ctx.path],
        expected: "string",
        received: describeType(input),
      });
      return undefined;
    }

    const val = input;
    const length = Array.from(val).length;

    if (this._minLength !== undefined && length < this._minLength) {
      ctx.issues.push({
        code: ISSUE_CODES.TOO_SMALL,
        message: `String must have at least ${this._minLength} character(s)`,
        path: [...ctx.path],
        expected: String(this._minLength),
        received: String(length),
      });
    }

    if (this._maxLength !== undefined && length > this._maxLength) {
      ctx.issues.push({
        code: ISSUE_CODES.TOO_LARGE,
        message: `String must have at most ${this._maxLength} character(s)`,
        path: [...ctx.path],
        expected: String(this._maxLength),
        received: String(length),
      });
    }

    if (this._pattern !== undefined) {
      const re = this._getPatternRe();
      if (re === null) {
        // Invalid regex pattern - treat as validation failure
        ctx.issues.push({
          code: ISSUE_CODES.INVALID_STRING,
          message: `Invalid regex pattern: ${this._pattern}`,
          path: [...ctx.path],
          expected: this._pattern,
          received: val,
        });
      } else if (!re.test(val)) {
        ctx.issues.push({
          code: ISSUE_CODES.INVALID_STRING,
          message: `String does not match pattern: ${this._pattern}`,
          path: [...ctx.path],
          expected: this._pattern,
          received: val,
        });
      }
    }

    if (this._startsWith !== undefined && !val.startsWith(this._startsWith)) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_STRING,
        message: `String must start with "${this._startsWith}"`,
        path: [...ctx.path],
        expected: this._startsWith,
        received: val,
      });
    }

    if (this._endsWith !== undefined && !val.endsWith(this._endsWith)) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_STRING,
        message: `String must end with "${this._endsWith}"`,
        path: [...ctx.path],
        expected: this._endsWith,
        received: val,
      });
    }

    if (this._includes !== undefined && !val.includes(this._includes)) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_STRING,
        message: `String must include "${this._includes}"`,
        path: [...ctx.path],
        expected: this._includes,
        received: val,
      });
    }

    if (this._format !== undefined && !validateFormat(val, this._format)) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_STRING,
        message: `Invalid ${this._format} format`,
        path: [...ctx.path],
        expected: this._format,
        received: val,
      });
    }

    return val;
  }

  _toNode(): SchemaNode {
    const node: Record<string, unknown> = { kind: "string" };
    if (this._minLength !== undefined) node.minLength = this._minLength;
    if (this._maxLength !== undefined) node.maxLength = this._maxLength;
    if (this._pattern !== undefined) node.pattern = this._pattern;
    if (this._startsWith !== undefined) node.startsWith = this._startsWith;
    if (this._endsWith !== undefined) node.endsWith = this._endsWith;
    if (this._includes !== undefined) node.includes = this._includes;
    if (this._format !== undefined) node.format = this._format;
    this._addDefault(node as unknown as SchemaNode);
    return node as unknown as SchemaNode;
  }
}
